File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int min (vector<int> arr)
{
	int c = 1;
	for(int i=0;i<arr.size();i++)
	{
		if(arr[i] < c) c= arr[i];
	}
	return c;

}
int max (vector<int> arr)
{
	int c = 1;
	for(int i=0;i<arr.size();i++)
	{
		if(arr[i] > c) c= arr[i];
	}
	return c;

}

int possible(vector<int> arr, int day, int k, int m)
{
	int count = 0;
	int ans =0;
	for(int i=0;i<arr.size();i++)
	{
		if(day>= arr[i]) count++;
		else {
			ans += count/k;
			count =0;
		}
	}
	ans += count/k;
	return ans;
}

    
    int minDays(vector<int>& arr, int m, int k) {
        long long s = min(arr);
	long long e = max(arr);

	long long ans = -1;

	while(s<=e)
	{
		long long mid = s + (e-s)/2;

		int output = possible( arr, mid, k,  m);
		// cout<<output<<endl;
		if(output >= m)
		{
			ans =mid;
			e = mid-1; 
		}
		else s = mid+1;

	}

	return ans;

    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public static int max(int[] b)
    {
        int m = 0;
        for(int i=0;i<b.length;i++)
        {
            if(b[i]>m) m = b[i];
        }
        return m;
    }
    
    public static int helper(int[] b , int mid,int k)
    {
        int temp = 0;
        int count = 0;
        for(int i=0;i<b.length;i++)
        {
            if(b[i] <= mid )
            {                
                temp++;
                if(temp == k){
                    count++;
                    temp = 0;
                }
            }
            else {
                temp = 0;
            }
        }
        
        return count;
    }
    
    public int minDays(int[] b, int m, int k) {
        int s = 1;
        int e = max(b);
        
        if(m*k > b.length) return -1;
        int ans = -1;
        
        while(s<=e)
        {
            int mid = s + (e-s)/2;
            int h = helper(b,mid,k);
            System.out.println(h);
            if(h >= m){
                ans = mid;
                e= mid-1;
            }
            else s = mid+1;
        }
        return ans;
    }
}